tests/test-sessions/test-sessions-common/src/main/java/org/eclipse/jetty/server/session/AbstractSessionDataStoreTest.java

```diff
@@ -205,37 +206,109 @@ public void testUpdateSession() throws Exception
+        File factoryClass = new File (proxyabledir, "ProxyableFactory.class");
+        IO.copy(is, new FileOutputStream(factoryClass));
+        is.close();
+
+        URL[] proxyabledirUrls = new URL[]{proxyabledir.toURI().toURL()};
+        _contextClassLoader = new URLClassLoader(proxyabledirUrls, Thread.currentThread().getContextClassLoader());
+        
+        //create the SessionDataStore 
+        ServletContextHandler context = new ServletContextHandler(ServletContextHandler.SESSIONS);
+        context.setContextPath("/test");
+        //use the classloader with the special class in it
+        context.setClassLoader(_contextClassLoader);
+        
+        SessionDataStoreFactory factory = createSessionDataStoreFactory();
+        ((AbstractSessionDataStoreFactory)factory).setGracePeriodSec(GRACE_PERIOD_SEC);
+        SessionDataStore store = factory.getSessionDataStore(context.getSessionHandler());
+        SessionContext sessionContext = new SessionContext("foo", context.getServletContext());
+        store.initialize(sessionContext);
+
+        store.start();
+        
+        ClassLoader old = Thread.currentThread().getContextClassLoader();
+        SessionData data = null;
+        try
+        {
+            //Set the classloader and make a session containing a proxy as an attribute
+            Thread.currentThread().setContextClassLoader(_contextClassLoader);
+            Class factoryclazz = Class.forName("ProxyableFactory", true, _contextClassLoader);
+            //create a session
+            long now = System.currentTimeMillis();
+            data = store.newSessionData("1234", 100, now, now-1, -1);//never expires
+            data.setLastNode(sessionContext.getWorkerName());
+            Method m = factoryclazz.getMethod("newProxyable", ClassLoader.class);
+            Object proxy = m.invoke(null, _contextClassLoader);
+
+            //Make an attribute that uses the proxy only known to the webapp classloader
+            data.setAttribute("a", proxy);
+        }
+        finally
+        {
+            Thread.currentThread().setContextClassLoader(old);
+        }
+        
+        
+        //store the session, using a different thread to ensure
+        //that the thread is adorned with the webapp classloader
+        //before serialization
+        final SessionData finalData = data;
+        
+        Runnable r = new Runnable()
+        {
+
+            @Override
+            public void run()
+            {
+                try
+                {
+                    store.store("1234", finalData);
+                }
+                catch (Exception e)
+                {
+                    fail(e);
+                }
+            }
+        };
+       
+        Thread t = new Thread(r, "saver");
+        t.start();
+        t.join(TimeUnit.SECONDS.toMillis(10));
+        
+        //check that the store contains all of the session data
+        assertTrue(checkSessionPersisted(data));
+    }
```

Comment:
Is there an equivalent method that tests that the proxy can be deserialized from the store?